gossip/privdata/dataretriever.go

```diff
@@ -55,14 +55,25 @@ func (dr *dataRetriever) CollectionRWSet(digests []*protosgossip.PvtDataDigest,
 		// if there is an error getting info from the ledger, we need to try to read from transient store
 		return nil, false, errors.Wrap(err, "wasn't able to read ledger height")
 	}
-	if height <= blockNum {
+
+	// The condition may be true for either commit or reconciliation case when another peer sends a request to retrieve private data.
+	// For the commit case, get the private data from the transient store because the block has not been committed.
+	// For the reconciliation case (when TxId is empty), it means this peer is further behind the ledger height than the peer
+	// that requested for the private data. Therefore, neither transient store nor ledger has the requested private data.
+	if height <= blockNum { // Check whenever current ledger height is equal or below block sequence num.
 		dr.logger.Debug("Current ledger height ", height, "is below requested block sequence number",
 			blockNum, "retrieving private data from transient store")
-	}
 
-	if height <= blockNum { // Check whenever current ledger height is equal or below block sequence num.
 		results := make(Dig2PvtRWSetWithConfig)
 		for _, dig := range digests {
+			// It is the reconciliation case since TxId is empty, skip retrieving private data from transient store.
+			if dig.TxId == "" {
```

Comment:
Can we formulate the condition differently? These hidden meanings attached to special values of fields (0, -1 for integers, zero-length for strings etc.) tend to be quickly forgotten, and then it's hard to find out what goes wrong in the future. It would be better if we could find a way of making it explicit that it's reconciliation without resorting to zero-length string check.

Reply:
Thank you Ale. Agree it would be nice to have a way to make it explicit. Ideally, the puller request message should contain a field to explicitly indicate the request type (e.g., reconciliation vs. commit). However, the protobuf message does not have such a field. Adding a new field will introduce backward compatibility issue where some peers are in new versions and others may be in old version. Therefore, this PR adds a new func isReconciliationRequest to help clarify the use case for reconciliation requests.

If you still think we need a field to explicitly indicate the request type, we can open a new jira since it is out of the scope of this PR for bug fix.

Reply:
Thanks @wenjianqiao , I think this is already better as it makes the fact at least explicit (with tests attached). LGTM